wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -305,11 +306,111 @@ private class WasmExpressionBuilder private (
           IRTrees.ApplyStatically(t.flags, t.receiver, className, t.method, t.args)(t.tpe)(t.pos)
         )
 
+      case IRTypes.AnyType if t.method.name.isReflectiveProxy =>
+        genReflectiveCall(t)
+
       case _ =>
         genApplyNonPrim(t)
     }
   }
 
+  private def genReflectiveCall(t: IRTrees.Apply): IRTypes.Type = {
+    // A local for a copy of the receiver that we will use to resolve dispatch
+    val receiverLocalForDispatch =
+      fctx.addSyntheticLocal(Types.WasmRefType.any)
+    val reflectiveProxies =
+      fctx.addSyntheticLocal(Types.WasmRefType(WasmArrayTypeName.reflectiveProxies))
+    val size = fctx.addSyntheticLocal(Types.WasmInt32)
+    val i = fctx.addSyntheticLocal(Types.WasmInt32)
+
+    // taking a hashCode for ease, but it might be better to compare in string
+    // to completely get rid of the risk of hash collision(?)
+    val methodName = t.method.name.nameString.hashCode()
+    // println(s"reflective call on ${t.method.name.nameString} ${methodName}")
+
+    val receiverType = TypeTransformer.makeReceiverType
+    val paramTys = t.method.name.paramTypeRefs.map(TypeTransformer.transformTypeRef(_)(ctx))
+    val sig = WasmFunctionSignature(
+      receiverType +: paramTys,
+      TypeTransformer.transformResultType(IRTypes.AnyType)(ctx)
+    )
+    val funcTy = WasmFunctionType(ctx.addFunctionType(sig), sig)
+
+    fctx.block(sig.results) { done =>
+      fctx.block(Types.WasmRefType.any) { labelNotOurObject =>
+        genTree(t.receiver, IRTypes.AnyType)
+        instrs += REF_AS_NOT_NULL
+        instrs += LOCAL_TEE(receiverLocalForDispatch)
+        instrs += BR_ON_CAST_FAIL(
+          labelNotOurObject,
+          Types.WasmRefType.any,
+          Types.WasmRefType(Types.WasmHeapType.ObjectType)
+        )
+
+        instrs += STRUCT_GET(
+          WasmStructTypeName.forClass(IRNames.ObjectClass),
+          WasmFieldIdx.vtable
+        )
+        instrs += STRUCT_GET(
+          WasmTypeName.WasmStructTypeName.typeData,
+          WasmFieldIdx.typeData.reflectiveProxiesIdx
+        )
+        instrs += LOCAL_TEE(reflectiveProxies)
+        instrs += ARRAY_LEN
+        instrs += LOCAL_SET(size)
+
+        instrs += I32_CONST(0)
+        instrs += LOCAL_SET(i)
+
+        fctx.whileLoop() {
```

Comment:
That's a big loop. I think it can be put in a helper function with the signature
```wat
[typeData, i32] -> [(ref func)]
```
Then at call site we only need, in pseudo-code
```wat
load vtable
i32.const <method ID>
call $theHelper
ref.cast Types.WasmRefType(Types.WasmHeapType(funcTy.name))
call_ref funcTy.name
```